### PR TITLE
feat: [GM-26] [BE] Implement the get tags endpoint

### PR DESCRIPTION
**Pull request by krishankumar01**
Example: 

image
image
image

### CODE REVIEW

File: src/proto/package/tag.proto [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+option go_package="market.goravel.dev/proto/package";
+
+import "google/api/annotations.proto";
+import "base/base.proto";
+
+message Tag {
+  string id = 1;
+  string user_id = 2;
+  string name = 3;
+  string description = 4;
```
Comment (hwbrzzl) [inline src/proto/package/tag.proto]:
```suggestion
```

File: src/proto/README.md
Comment (hwbrzzl) [inline src/proto/README.md:1]: Changed this in your another PR, Why is this here?

Comment (krishankumar01) [reply src/proto/README.md:1]: removed

File: src/go/gateway/config/grpc.go [hunk 2]
```diff
@@ -25,6 +26,14 @@ func init() {
...
 			},
+			"package": map[string]any{
+				"host": config.Env("GRPC_PACKAGE_HOST", ""),
+				"port": config.Env("GRPC_PACKAGE_PORT", ""),
+				"handlers": []gateway.Handler{
+					protopackage.RegisterPackageServiceHandler,
+				},
+				"interceptors": []string{},
+			},
 			//"business": map[string]any{
```
Comment (hwbrzzl) [inline src/go/gateway/config/grpc.go]: These lines can be removed.

File: src/go/gateway/routes/package.go [hunk 3]
```diff
@@ -0,0 +1,10 @@
+package routes
+
+import (
+	"github.com/goravel/framework/facades"
+	"github.com/goravel/gateway"
+)
+
+func Packages() {
+	facades.Route().Get("/tags", gateway.Get)
```
Comment (hwbrzzl) [inline src/go/gateway/routes/package.go]:
```suggestion
	facades.Route().Get("/packages/tags", gateway.Get)
```

File: src/go/proto/package/tag.pb.gw.go
Comment (hwbrzzl) [inline src/go/proto/package/tag.pb.gw.go:1]: Why remove this file?

File: src/go/proto/package/tag_grpc.pb.go
Comment (hwbrzzl) [inline src/go/proto/package/tag_grpc.pb.go:1]: Ditto

File: src/go/package/database/migrations/20240101035305_create_tags_table.up.sql [hunk 4]
```diff
@@ -1,6 +1,10 @@
 CREATE TABLE tags (
   id bigint PRIMARY KEY,
+  user_id bigint NOT NULL,
```
Comment (hwbrzzl) [inline src/go/package/database/migrations/20240101035305_create_tags_table.up.sql]: Can be NULL when imported by system.

File: src/go/package/database/migrations/20240101035305_create_tags_table.up.sql [hunk 5]
```diff
@@ -1,6 +1,10 @@
...
   name varchar(255) NOT NULL,
+  description text DEFAULT NULL,
+  is_show int DEFAULT 2,
```
Comment (hwbrzzl) [inline src/go/package/database/migrations/20240101035305_create_tags_table.up.sql]: Default 1, and add a comment: 1 show, 2 not show.

File: src/go/package/app/grpc/controllers/package_controller.go [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+}
+
+func NewPackageController() *PackageController {
+	return &PackageController{
+		tagService: services.NewTagImpl(),
+	}
+}
+
+func (r *PackageController) GetTags(ctx context.Context, req *protopackage.GetTagsRequest) (*protopackage.GetTagsResponse, error) {
+	query := req.GetQuery()
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller.go:24]: Need to judge if `query` is nil.

File: src/go/package/app/services/tag.go [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+
+import (
+	"github.com/goravel/framework/facades"
+
+	"market.goravel.dev/package/app/models"
+	protobase "market.goravel.dev/proto/base"
+)
+
+type Tag interface {
+	GetTags(packageID, name string, pagination *protobase.Pagination, total *int64) ([]*models.Tag, error)
```
Comment (hwbrzzl) [inline src/go/package/app/services/tag.go]:
```suggestion
	GetTags(packageID, name string, pagination *protobase.Pagination) ([]*models.Tag, int64, error)
```

File: src/go/package/app/grpc/controllers/package_controller.go [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+	tags, err := r.tagService.GetTags(packageID, name, pagination, &total)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(tags) == 0 {
+		return &protopackage.GetTagsResponse{
+			Status: utilsresponse.NewNotFoundStatus(errors.New(facades.Lang(ctx).Get("not_exist.tags"))),
+		}, nil
+	}
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller.go]: Returns empty tags.

File: src/go/package/app/grpc/controllers/package_controller_test.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+		userID     = "1"
+		name       = "go"
+		pagination = &protobase.Pagination{
+			Page:  1,
+			Limit: 10,
+		}
+		total int64
+	)
+
+	tests := []struct {
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller_test.go:54]: Add two test cases that query is nil, or pagination is nil.

File: src/go/package/app/grpc/controllers/package_controller.go [hunk 10]
```diff
@@ -0,0 +1,53 @@
...
+	return &PackageController{
+		tagService: services.NewTagImpl(),
+	}
+}
+
+func (r *PackageController) GetTags(ctx context.Context, req *protopackage.GetTagsRequest) (*protopackage.GetTagsResponse, error) {
+	query := req.GetQuery()
+	packageID := query.GetPackageId()
+	name := query.GetName()
+	pagination := req.GetPagination()
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller.go:27]: Need to judge pagination is not nil, and pagination.page, pagination.limit have a valid value.

File: src/go/package/app/models/tag.go [hunk 11]
```diff
@@ -2,10 +2,37 @@ package models
...
+
 type Tag struct {
-	orm.Model
-	Name string
+	UUIDModel
+	UserID      uint64
+	Name        string
+	Description string
+	IsShow      uint
+	Packages    []*Package `gorm:"many2many:package_tags;"`
```
Comment (hwbrzzl) [inline src/go/package/app/models/tag.go]: Don't need this for now.

File: src/go/package/app/models/tag_test.go [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+
+func TestTagSuite(t *testing.T) {
+	suite.Run(t, new(TagSuite))
+}
+
+func (s *TagSuite) SetupTest() {
+	s.tag = NewTag()
+}
+
+func (s *TagSuite) TestToProto() {
```
Comment (hwbrzzl) [inline src/go/package/app/models/tag_test.go]: Good habits, but the function is very simple, don't need to test it.

File: src/go/package/app/services/tag.go [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+		tagModel: models.NewTag(),
+	}
+}
+
+func (r *TagImpl) GetTags(packageID, name string, pagination *protobase.Pagination, total *int64) ([]*models.Tag, error) {
+	var tags []*models.Tag
+	query := facades.Orm().Query()
+
+	if packageID != "" {
+		query = query.Join("JOIN package_tags ON package_tags.tag_id = tags.id").
```
Comment (hwbrzzl) [inline src/go/package/app/services/tag.go]: Don't need to use the join sql, if packageID != "", we can get  the tagIDs from the package_tags through packageID: split to two sqls. `Json` is inefficiency, split is better.

**Review by hwbrzzl — COMMENTED** [review]
Great job 👍

File: src/go/proto/package/tag.pb.gw.go
Comment (krishankumar01) [reply src/go/proto/package/tag.pb.gw.go:1]: After removing tag service when I re run these file wasn't generated. (Chrck by removing original files)

File: src/go/package/app/models/tag_test.go (hunk 12 above)
Comment (krishankumar01) [reply src/go/package/app/models/tag_test.go]: removed it

File: src/go/utils/pagination/pagination.go [hunk 14]
```diff
@@ -0,0 +1,10 @@
+package pagination
+
+import protobase "market.goravel.dev/proto/base"
+
+func Default() *protobase.Pagination {
```
Comment (hwbrzzl) [inline src/go/utils/pagination/pagination.go:5]: 👍

File: src/go/package/database/migrations/20240101035305_create_tags_table.up.sql [hunk 15]
```diff
@@ -1,6 +1,11 @@
 CREATE TABLE tags (
   id bigint PRIMARY KEY,
+  user_id bigint DEFAULT NULL,
   name varchar(255) NOT NULL,
+  description text DEFAULT NULL,
+  /* 1: show, 2: not show */
```
Comment (hwbrzzl) [inline src/go/package/database/migrations/20240101035305_create_tags_table.up.sql]: Use a sql instead: `comment on column tags.is_show is '1: show, 2: not show'`

File: src/go/package/app/grpc/controllers/package_controller_test.go [hunk 16]
```diff
@@ -0,0 +1,183 @@
...
+	mockstranslation "github.com/goravel/framework/mocks/translation"
+	testingmock "github.com/goravel/framework/testing/mock"
+	"github.com/stretchr/testify/suite"
+
+	"market.goravel.dev/package/app/models"
+	protobase "market.goravel.dev/proto/base"
+	protopackage "market.goravel.dev/proto/package"
+	utilsresponse "market.goravel.dev/utils/response"
+
+	mocksservice "market.goravel.dev/package/app/mocks/services"
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller_test.go]:
```suggestion
	utilsresponse "market.goravel.dev/utils/response"
	mocksservice "market.goravel.dev/package/app/mocks/services"
```

File: src/go/package/app/grpc/controllers/package_controller_test.go [hunk 17]
```diff
@@ -0,0 +1,183 @@
...
+				},
+			},
+			setup: func() {
+				total = 0
+				s.mockTagService.On("GetTags", "", name, pagination).Return(nil, total, errors.New("error")).Once()
+			},
+			expectedErr: errors.New("error"),
+		},
+		{
+			name: "Sad path - tags is empty",
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller_test.go]:
```suggestion
			name: "Happy path - tags is empty",
```

File: src/go/package/app/grpc/controllers/package_controller_test.go [hunk 18]
```diff
@@ -0,0 +1,183 @@
...
+				total = 0
+				s.mockTagService.On("GetTags", "", name, pagination).Return([]*models.Tag{}, total, nil).Once()
+			},
+			expectedResponse: &protopackage.GetTagsResponse{
+				Status: utilsresponse.NewOkStatus(),
+				Tags:   []*protopackage.Tag{},
+			},
+		},
+		{
+			name: "Sad path - pagination is nil",
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller_test.go]:
```suggestion
			name: "Happy path - pagination is nil",
```

File: src/go/package/app/services/tag.go [hunk 19]
```diff
@@ -0,0 +1,50 @@
...
+func (r *TagImpl) GetTags(packageID, name string, pagination *protobase.Pagination) ([]*models.Tag, int64, error) {
+	var tags []*models.Tag
+	query := facades.Orm().Query()
+	var total int64
+
+	page := pagination.GetPage()
+	limit := pagination.GetLimit()
+
+	if packageID != "" {
+		var tagIDs []any
```
Comment (hwbrzzl) [inline src/go/package/app/services/tag.go:30]: Use a specific type instead of `any`.

File: src/go/package/app/services/tag.go [hunk 20]
```diff
@@ -0,0 +1,50 @@
...
+		}
+
+		query = query.WhereIn("id", tagIDs)
+	}
+
+	if name != "" {
+		// fuzzy search
+		query = query.Where("name LIKE ?", "%"+name+"%")
+	}
+	if err := query.Select([]string{"id", "name", "user_id"}).Paginate(int(page), int(limit), &tags, &total); err != nil {
```
Comment (hwbrzzl) [inline src/go/package/app/services/tag.go]:
```suggestion
	if err := query.Select([]string{"id", "name"}).Paginate(int(page), int(limit), &tags, &total); err != nil {
```

File: src/go/package/app/services/tag_test.go [hunk 21]
```diff
@@ -0,0 +1,30 @@
...
+
+func (s *TagTestSuite) SetupTest() {
+	s.mockTag = new(mocksmodels.TagInterface)
+	s.tagImpl = &TagImpl{
+		tagModel: s.mockTag,
+	}
+}
+
+func (s *TagTestSuite) TestGetTags() {
+	// TODO: Implement me
```
Comment (hwbrzzl) [inline src/go/package/app/services/tag_test.go]: Implement it.

File: src/go/package/app/models/tag.go [hunk 22]
```diff
@@ -2,10 +2,36 @@ package models
...
 import (
 	"github.com/goravel/framework/database/orm"
+	"github.com/goravel/framework/support/carbon"
+	"github.com/spf13/cast"
+
+	protopackage "market.goravel.dev/proto/package"
 )
 
+type TagInterface interface {
+	//GetTags(packageID, userID, name string, fields []string) ([]*Tag, error)
```
Comment (hwbrzzl) [inline src/go/package/app/models/tag.go]: Remove this if not needed.

File: src/go/package/app/models/tag_test.go
Comment (hwbrzzl) [inline src/go/package/app/models/tag_test.go:1]: Can be removed.

File: src/go/package/app/grpc/controllers/package_controller_test.go [hunk 23]
```diff
@@ -0,0 +1,182 @@
...
+			},
+			setup: func() {
+				total = 1
+				s.mockTagService.On("GetTags", packageID, name, pagination).Return([]*models.Tag{
+					{
+						UUIDModel: models.UUIDModel{
+							ID: 1,
+						},
+						UserID: 1,
+						Name:   "goravel",
```
Comment (hwbrzzl) [inline src/go/package/app/grpc/controllers/package_controller_test.go]:
```suggestion
						UserID: userID,
						Name:   name,
```

File: src/go/package/app/services/tag_test.go [hunk 24]
```diff
@@ -0,0 +1,206 @@
...
+				mockOrmQuery.On("Where", "is_show = ?", "1").Return(mockOrmQuery).Once()
+				mockOrmQuery.On("Paginate", int(pagination.GetPage()), int(pagination.GetLimit()), mock.AnythingOfType("*[]*models.Tag"), mock.AnythingOfType("*int64")).
+					Return(nil).
+					Run(func(args mock.Arguments) {
+						tagsPtr := args.Get(2).(*[]*models.Tag)
+						*tagsPtr = []*models.Tag{{UUIDModel: models.UUIDModel{ID: 1}, Name: "GoLang"}, {UUIDModel: models.UUIDModel{ID: 2}, Name: "Gopher"}}
+
+						totalPtr := args.Get(3).(*int64)
+						*totalPtr = 2
+					}).Once()
```
Comment (hwbrzzl) [inline src/go/package/app/services/tag_test.go:79]: 👍

File: src/go/package/database/migrations/20240101035305_create_tags_table.up.sql [hunk 25]
```diff
@@ -1,6 +1,10 @@
 CREATE TABLE tags (
   id bigint PRIMARY KEY,
+  user_id bigint DEFAULT NULL,
   name varchar(255) NOT NULL,
+  description text DEFAULT NULL,
```
Comment (hwbrzzl) [inline src/go/package/database/migrations/20240101035305_create_tags_table.up.sql]:
```suggestion
```

File: src/go/package/database/migrations/20240101035305_create_tags_table.up.sql [hunk 26]
```diff
@@ -1,6 +1,10 @@
...
+  is_show int DEFAULT 1 COMMENT '1: show, 2: not show',
```
Comment (hwbrzzl) [inline src/go/package/database/migrations/20240101035305_create_tags_table.up.sql]: Should like below:

```
CREATE TABLE employees (
    employee_id SERIAL PRIMARY KEY,
    first_name VARCHAR(50),
    last_name VARCHAR(50),
    hire_date DATE
);

COMMENT ON COLUMN employees.first_name IS 'The first name of the employee.';
COMMENT ON COLUMN employees.last_name IS 'The last name of the employee.';
COMMENT ON COLUMN employees.hire_date IS 'The date when the employee was hired.';
```

**Comment by hwbrzzl** [comment]
I saw the endpoint returns `user_id: "0"`, it's not good, the endpoint shouldn't return the user_id field, even if `user_id != ""`. So I think we'd better define a response struct for this endpoint, don't use `ToProto` directly.

**Review by hwbrzzl — APPROVED** [review]
Great
